cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -460,17 +469,21 @@ struct StructFieldFunctor {
   static Status CheckIndex(int index, const DataType& type) {
     if (!ValidParentType(type)) {
       return Status::TypeError("struct_field: cannot subscript field of type ", type);
-    } else if (index < 0 || index >= type.num_fields()) {
+    } else if (!IsBaseListType(type) && (index < 0 || index >= type.num_fields())) {
```

Comment:
In theory, we could check the index for fixed sized list arrays as well (not for variable size list arrays, though, so not sure that is worth it)